designs/0013-profiling.md

```diff
@@ -0,0 +1,231 @@
+- *Feature Name:* profiling-stan-AD
+- *Start Date:* NA
+- *RFC PR(S):*
+- *Stan Issue(s):*
+
+# Summary
+[summary]: #summary
+
+The goal of this design is a clean and simple interface to profiling Stan models. Specfically, profiling single nodes and subexpressions of the autodiff (AD) expression graph for the model and transformed parameters blocks and function evaluation for the generated quantities and transformed data blocks.
+
+# Motivation
+[motivation]: #motivation
+
+1. identifying bottlenecks in a Stan model
+
+Currently, identifying bottlenecks requires a good understanding of both autodiff and its implementation in Stan Math. And even with that knowledge, its not feasible to confirm an assumed bottleneck is a true bottleneck. Optimization of a Stan model for faster per-gradient execution is therefore very difficult.
+
+2. evaluating different approaches for a section of a model
+
+It is very difficult to evaluate how a different approach to a secton of a model influences per-gradient performance as even small numerical differences can lead to different behaviour of the sampler and the entire model must to be run with different seeds for a fair comparison. And even then, the effect of the changes can be difficult to identify in the overall noise of a model execution that includes IO and other parts of a model.
+
+For these cases, profiling a subexpression in the AD expression graph is required. For some functions it can still be seed-dependent, but to a much lesser extent especially with increasing numbers of gradient/function evaluations. 
+
+Examples of cases where evaluating is very useful: evaluating paralleliziation approaches (reduce_sum, map_rect, GPUs), identifying the effect of using a different solver or a different expression.
+
+# Profiling on the Stan model level
+[stan-model-profiling]: #stan-model-profiling
+
+A user defines profiling sections inside a model by writing `profile("profile-section-name");`. The argument to `profile()` is always a string literal. A profile will measure execution time and other AD information for all expressions for the Stan lines for which the profile is in scope.
+
+An model with examples of its usage is shown below:
+
+```stan
```

Comment:
I would like this broken into a smaller complete example before doing a monolithic model with all possible uses.